### PR TITLE
Minhdev

### PR DESCRIPTION
**Pull request by hoanghongminh**
refactor code, hook >< UI, create component common, add router and HOC withAuth, create pagination, login authentication

### CODE REVIEW

File: src/components/leaderboard/LeaderboardTable.tsx [hunk 1]
```diff
@@ -81,6 +81,9 @@ export const LeaderboardTable = ({ data }: Props) => {
           totalPages={2}
           onClickNextPage={handleNextPage}
           onClickPrevPage={handlePrevPage}
+          onPageClick={function (pageNumber: number): void {
+            throw new Error("Function not implemented.");
```
Comment (nxthai23) [inline src/components/leaderboard/LeaderboardTable.tsx:85]: why throw error, should we set tooltip like coming soon or disable?

File: src/components/ui/pagination.tsx
Comment (nxthai23) [inline src/components/ui/pagination.tsx:1]: remove if not using

File: src/lib/hooks/useEvent.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+const API_URL = "http://localhost:3500/events";
+
+export const useEvents = () => {
+  const [events, setEvents] = useState<Event[]>([]);
+
+  useEffect(() => {
+    const fetchEvents = async () => {
+      try {
+        const response = await fetch(API_URL);
```
Comment (nxthai23) [inline src/lib/hooks/useEvent.ts:13]: use axios here

File: src/lib/hooks/useEventDetail.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    month: "",
+    day: "",
+    weekday: "",
+    content: "",
+  });
+
+  useEffect(() => {
+    const fetchEventDetail = async () => {
+      try {
+        const response = await fetch(API_URL);
```
Comment (nxthai23) [inline src/lib/hooks/useEventDetail.ts:22]: use axios

File: src/mock/news-mock.json
Comment (nxthai23) [inline src/mock/news-mock.json:1]: move all mock.json to db.json => then use axios to get data from json-server first

File: src/requests/login.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+type LoginBody = {
+  identifier: string;
+  password: string;
+  role: {
+    name: string;
+  };
+};
+
+// Backend base url
+const BASE_URL = "http://localhost:1337/api";
```
Comment (nxthai23) [inline src/requests/login.ts:12]: move to const

File: src/components/hoc/WithAuth.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import { useUserStore } from "@/store/UserStore";
+import { toast, ToastContainer } from "react-toastify";
+
+const WithAuth = (WrappedComponent: React.FC) => {
+  const Comp: React.FC = () => {
+    const router = useRouter();
+    const [jwt] = useUserStore((state) => [state.jwt]);
+
+    useEffect(() => {
+      if (!jwt) {
```
Comment (nxthai23) [inline src/components/hoc/WithAuth.tsx:13]: use isConnected() in userStore

File: data/eventDetail.json
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply data/eventDetail.json:1]: done

File: data/events.json
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply data/events.json:1]: done

File: src/app/event/[id]/page.tsx [hunk 6]
```diff
@@ -1,27 +1,29 @@
 "use client";
+import React from "react";
 import { ArrowLeft, Clock4, MapPin } from "lucide-react";
-import { useRouter } from "next/navigation";
-import React, { useState } from "react";
+import { useRouter } from "next/router";
+// import { EventDetail } from "@/types/common-types";
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/event/[id]/page.tsx:5]: done

File: src/app/event/layout.tsx [hunk 7]
```diff
@@ -25,7 +25,6 @@ export default function Layout({
   const [userName, setUserName] = useState("HENRY NGUYEN");
   const [userRank, setUserRank] = useState("BẠC IV");
   const [show, hide] = useProfileStore((state) => [state.show, state.hide]);
-  const router = useRouter();
   const handleOpenModal = () => {
     show("aaaa");
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/event/layout.tsx:29]: fixed => content from Actions show(content) ProfileStore

File: src/app/home/layout.tsx [hunk 8]
```diff
@@ -28,11 +22,11 @@ export default function Layout({
   const [userName, setUserName] = useState("HENRY NGUYEN");
   const [userRank, setUserRank] = useState("BẠC IV");
   const [show, hide] = useProfileStore((state) => [state.show, state.hide]);
-  const router = useRouter();
   const handleOpenModal = () => {
     show("aaaa");
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/home/layout.tsx:26]: fixed

File: src/app/event/layout.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/event/layout.tsx:29]: remove because not use now

File: src/app/home/layout.tsx [hunk 9]
```diff
@@ -28,11 +22,11 @@ export default function Layout({
...
   };
-  const projectsSliced = projectsMock.slice(1, 5);
+  const projectsSliced = projectsMock.slice(1, 5) as Project[];
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/home/layout.tsx:28]: fixed: projectsSliced => projectsList

File: src/app/home/profile/page.tsx [hunk 10]
```diff
@@ -24,19 +24,33 @@ import { Post } from "@/components/home/Post";
...
   ]);
 
+  const getMe = useUserStore((state) => state.getMe);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const userData = await getMe();
+      } catch (error) {
+        console.error("Error fetching user data:", error);
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/home/profile/page.tsx:48]: fixed

File: src/app/mission/layout.tsx [hunk 11]
```diff
@@ -28,11 +26,11 @@ export default function Layout({
   const [userName, setUserName] = useState("HENRY NGUYEN");
   const [userRank, setUserRank] = useState("BẠC IV");
   const [show, hide] = useProfileStore((state) => [state.show, state.hide]);
-  const router = useRouter();
   const handleOpenModal = () => {
     show("aaaa");
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/mission/layout.tsx:30]: fixed

File: src/app/mission/layout.tsx [hunk 12]
```diff
@@ -28,11 +26,11 @@ export default function Layout({
...
   };
   const projectsSliced = projectsMock.slice(1, 5);
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/mission/layout.tsx:32]: fixed, rename projectsList

File: src/app/news/layout.tsx [hunk 13]
```diff
@@ -30,6 +22,7 @@ export default function Layout({
   const handleOpenModal = () => {
     show("aaaa");
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/news/layout.tsx:23]: fixed

File: src/lib/hooks/useEvent.ts (hunk 2 above)
Comment (hoanghongminh) [reply src/lib/hooks/useEvent.ts:13]: fixed

File: src/lib/hooks/useEventDetail.ts (hunk 3 above)
Comment (hoanghongminh) [reply src/lib/hooks/useEventDetail.ts:22]: fixed

File: src/requests/login.ts (hunk 4 above)
Comment (hoanghongminh) [reply src/requests/login.ts:12]: fixed

File: src/components/ui/pagination.tsx
Comment (hoanghongminh) [reply src/components/ui/pagination.tsx:1]: fixed

File: src/components/leaderboard/LeaderboardTable.tsx (hunk 1 above)
Comment (hoanghongminh) [reply src/components/leaderboard/LeaderboardTable.tsx:85]: fixed

File: src/components/hoc/WithAuth.tsx (hunk 5 above)
Comment (hoanghongminh) [reply src/components/hoc/WithAuth.tsx:13]: fixed

File: src/components/common/UserProfileLink.tsx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  userRank,
+}) => {
+  const router = useRouter();
+  const [isMenuOpen, setIsMenuShow] = useState(false);
+  const toggleMenu = () => {
+    setIsMenuShow(!isMenuOpen);
+  };
+  const handleLogout = () => {
+    router.push("/login");
+    useUserStore.getState().clear();
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/components/common/UserProfileLink.tsx:23]: yes, fixed

File: src/components/common/UserProfileLink.tsx [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  userName: string;
+  userRank: string;
+};
+
+const UserProfileLink: React.FC<UserProfileLinkProps> = ({
+  userName,
+  userRank,
+}) => {
+  const router = useRouter();
+  const [isMenuOpen, setIsMenuShow] = useState(false);
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/components/common/UserProfileLink.tsx:17]: fixed

File: src/app/recruitment/page.tsx [hunk 16]
```diff
@@ -1,15 +1,61 @@
...
+        const data = await response.json();
+        const startIndex = (currentPage - 1) * itemsPerPage;
+        const endIndex = startIndex + itemsPerPage;
+        const currentRecruitments = data.slice(startIndex, endIndex);
+        setRecruitments(currentRecruitments);
+
+        const totalPages = Math.ceil(data.length / itemsPerPage);
+        setTotalPages(totalPages);
+      } catch (error) {
+        console.error("Error fetching data:", error);
```
[the remark this answers is not in the document]
Comment (hoanghongminh) [reply src/app/recruitment/page.tsx:53]: fixed

**Pull request merged**